src/Chapter2.hs

```diff
@@ -858,7 +876,11 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
+rewind :: [a] -> [a]
+rewind [] = []
+rewind xs = lastElement : rewind (init xs)
+    where
+       lastElement = xs !! ((length xs) - 1)
```

Comment:
Your solution is correct! However, it is slow 🐌 
 In lists, it is quite slow to add anything at the end of the list, take at the position, or take the length. That is why it is always better to, for example, use `:` cons instead of `++`.  Remember the explanation with trains? 🚂 🚋 🚋 

That is why a more efficient solution is with the accumulator and the recursive function that will do the addition at the start of the list which is instant!

You can read a bit more about the `go` pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#recursive-go

Reply:
Thank you for the insights! my solutions tend to be a little bit complex than they need. 

Regarding the recursive-go pattern, I already saw that but it was called tail recursion, like this: https://youtu.be/_JtPhF8MshA

 Anyway, I fail to see how to apply tail-recursion/recursive go pattern to the rewind function. I mean, in the end I'll need to access to the last element anyway. I don't understand how tail-recursion can mitigate/avoid this issue.

Can you kindly show me that? Thanks for your awesome effort!!!

Reply:
Sure! I will try to explain with the reasoning at each step 🙂 

Taking element by index in lists in Haskell is not a cheap operation, due to the way lists are implemented. You need to go through each element until you meet the required position. That is why the solution with `!!` would not be efficient as well.

In Haskell, all such problems in lists are solved with the `:` cons and pattern matching.
Instead of the index, let's have as an accumulator the new list that we will construct on each recursive step.

```haskell
rewind :: [Int] -> [Int]
rewind l = go [] l
  where
    go :: [Int] -> [Int] -> [Int]
    go acc l
```

and now the most interesting begins.
Let's start our recursive function with the stop case. As we will go element by element, our final stop will be the empty list. In that case, we assume that the accumulator already has all elements reversed, so we return it:

```haskell
go acc [] = acc
```

Let's now look at the general case. Now it would be the mind-blowing moment (at least it was for me 🤯 ):

```haskell
go acc (x:xs) = go (x:acc) xs
```

So what we are doing? We are adding the elements to the start of the accumulator.

Let's prove that it will be reversed in the accumulator (even though we add to the beginning) on a small example:

| acc | list |
|-----|----|
| [] | [1,2,3,4]|
| 1: [] | [2,3,4] |
| 2 : 1 : [] | [3,4] |
| 3 : 2 : 1 : [] | [4] |
| 4 : 3 : 2 : 1 : [] | [] |

So at the last step, the list is reversed, and the function works only by one walk-through the list!

Hope it makes sense, let me know if you have questions 🙂

Reply:
Ah yes! Now I see how it works, brilliant. Thank you for the crystal-clear explanation!